File: Traditional-Important-Algorithms/exercises_py/06_distributed_systems/dist10_two_phase_commit.py
# I AM NOT DONE

"""
dist10_two_phase_commit.py

Two-Phase Commit (2PC) is a distributed algorithm that ensures all participants
in a distributed transaction either commit or abort together. It's crucial for
maintaining ACID properties across distributed databases.

Protocol phases:
1. Prepare Phase: Coordinator asks all participants to prepare (vote yes/no)
2. Commit Phase: If all vote yes, coordinator tells all to commit; otherwise abort

Your task: Implement the Two-Phase Commit protocol with coordinator and participants.

Key concepts:
- Coordinator: Orchestrates the transaction
- Participants: Nodes that execute parts of the transaction
- Prepare: Participant locks resources and votes
- Commit/Abort: Final decision applied atomically
- Blocking: Protocol blocks if coordinator fails (limitation of 2PC)
"""

from dataclasses import dataclass
from enum import Enum, auto
from typing import Dict, List, Optional
import unittest


class Vote(Enum):
    """Vote types."""
    YES = auto()
    NO = auto()


class ParticipantState(Enum):
    """Participant states."""
    IDLE = auto()
    PREPARED = auto()
    COMMITTED = auto()
    ABORTED = auto()


class CoordinatorState(Enum):
    """Coordinator states."""
    IDLE = auto()
    PREPARING = auto()
    COMMITTING = auto()
    ABORTING = auto()
    COMMITTED = auto()
    ABORTED = auto()


class TransactionResult(Enum):
    """Transaction result."""
    COMMITTED = auto()
    ABORTED = auto()
    IN_PROGRESS = auto()


@dataclass
class Transaction:
    """A distributed transaction."""
    id: int
    operations: List[str]


class Participant:
    """A participant in 2PC."""

    def __init__(self, participant_id: int):
        """Initialize a participant."""
        self.id = participant_id
        self.state = ParticipantState.IDLE
        self.prepared_transactions: Dict[int, Transaction] = {}
        self.committed_transactions: Dict[int, Transaction] = {}
        self.vote_to_abort = False  # Simulates failures

    def prepare(self, transaction: Transaction) -> Vote:
        """
        TODO: Handle prepare request.

        - If vote_to_abort is True, return Vote.NO
        - Otherwise, store transaction in prepared_transactions
        - Change state to PREPARED
        - Return Vote.YES
        """
        pass  # TODO: Implement this

    def commit(self, transaction_id: int) -> bool:
        """
        TODO: Handle commit request.

        - Must be in PREPARED state
        - Move transaction from prepared_transactions to committed_transactions
        - Change state to COMMITTED
        - Return True if successful, False otherwise
        """
        pass  # TODO: Implement this

    def abort(self, transaction_id: int) -> bool:
        """
        TODO: Handle abort request.

        - Remove transaction from prepared_transactions
        - Change state to ABORTED
        - Return True
        """
        pass  # TODO: Implement this

    def set_vote_to_abort(self, should_abort: bool):
        """Set whether this participant will vote to abort."""
        self.vote_to_abort = should_abort

    def is_transaction_committed(self, transaction_id: int) -> bool:
        """Check if a transaction is committed."""
        return transaction_id in self.committed_transactions

    def reset(self):
        """Reset to idle state."""
        self.state = ParticipantState.IDLE


class Coordinator:
    """The coordinator in 2PC."""

    def __init__(self):
        """Initialize a coordinator."""
        self.state = CoordinatorState.IDLE
        self.transaction: Optional[Transaction] = None
        self.votes: Dict[int, Vote] = {}
        self.participants: List[int] = []

    def add_participant(self, participant_id: int):
        """Add a participant."""
        self.participants.append(participant_id)

    def start_transaction(self, transaction: Transaction):
        """
        TODO: Start a new transaction.

        - Store the transaction
        - Clear votes
        - Change state to PREPARING
        """
        pass  # TODO: Implement this

    def receive_vote(self, participant_id: int, vote: Vote):
        """
        TODO: Record a vote from a participant.

        - Store the vote in votes dict
        """
        pass  # TODO: Implement this

    def has_all_votes(self) -> bool:
        """
        TODO: Check if all participants have voted.

        - Compare len(votes) with len(participants)
        """
        pass  # TODO: Implement this

    def should_commit(self) -> bool:
        """
        TODO: Determine if transaction should commit.

        - All votes must be YES
        - Return False if any vote is NO or not all votes received
        """
        pass  # TODO: Implement this

    def decide(self) -> TransactionResult:
        """
        TODO: Make commit/abort decision.

        - Must have all votes
        - If should_commit(), change state to COMMITTING, return COMMITTED
        - Otherwise, change state to ABORTING, return ABORTED
        - If not all votes, return IN_PROGRESS
        """
        pass  # TODO: Implement this

    def finalize_commit(self):
        """Finalize the commit."""
        self.state = CoordinatorState.COMMITTED

    def finalize_abort(self):
        """Finalize the abort."""
        self.state = CoordinatorState.ABORTED

    def reset(self):
        """Reset to idle state."""
        self.state = CoordinatorState.IDLE
        self.transaction = None
        self.votes.clear()


class TwoPhaseCommitSystem:
    """A complete 2PC system."""

    def __init__(self):
        """Initialize the system."""
        self.coordinator = Coordinator()
        self.participants: Dict[int, Participant] = {}
        self.transaction_counter = 0

    def add_participant(self, participant_id: int):
        """Add a participant to the system."""
        self.participants[participant_id] = Participant(participant_id)
        self.coordinator.add_participant(participant_id)

    def set_participant_vote(self, participant_id: int, vote_to_abort: bool):
        """Set whether a participant will vote to abort."""
        participant = self.participants.get(participant_id)
        if participant:
            participant.set_vote_to_abort(vote_to_abort)

    def execute_transaction(self, operations: List[str]) -> TransactionResult:
        """
        TODO: Execute a complete 2PC transaction.

        Phase 1: Prepare
        - Create transaction with next transaction_id
        - Start transaction on coordinator
        - Send prepare to all participants
        - Collect votes and send to coordinator

        Phase 2: Commit or Abort
        - Coordinator decides based on votes
        - If COMMITTED: send commit to all participants
        - If ABORTED: send abort to all participants
        - Finalize coordinator state
        - Return result
        """
        pass  # TODO: Implement this

    def phase1_prepare(self, transaction: Transaction) -> Dict[int, Vote]:
        """
        TODO: Execute Phase 1 (Prepare).

        - For each participant, call prepare
        - Collect and return all votes
        """
        pass  # TODO: Implement this

    def phase2_commit(self, transaction_id: int):
        """
        TODO: Execute Phase 2 Commit.

        - For each participant, call commit
        - Finalize coordinator
        """
        pass  # TODO: Implement this

    def phase2_abort(self, transaction_id: int):
        """
        TODO: Execute Phase 2 Abort.

        - For each participant, call abort
        - Finalize coordinator
        """
        pass  # TODO: Implement this

    def get_coordinator_state(self) -> CoordinatorState:
        """Get coordinator state."""
        return self.coordinator.state

    def get_participant_state(self, participant_id: int) -> Optional[ParticipantState]:
        """Get participant state."""
        participant = self.participants.get(participant_id)
        return participant.state if participant else None

    def is_transaction_committed(self, transaction_id: int) -> bool:
        """Check if all participants have committed the transaction."""
        return all(
            p.is_transaction_committed(transaction_id)
            for p in self.participants.values()
        )

    def reset_all(self):
        """Reset all components."""
        self.coordinator.reset()
        for participant in self.participants.values():
            participant.reset()


# Unit Tests
class TestTwoPhaseCommit(unittest.TestCase):

    def test_participant_prepare_yes(self):
        participant = Participant(1)
        tx = Transaction(1, ["INSERT INTO table"])

        vote = participant.prepare(tx)

        self.assertEqual(vote, Vote.YES)
        self.assertEqual(participant.state, ParticipantState.PREPARED)
        self.assertIn(1, participant.prepared_transactions)

    def test_participant_prepare_no(self):
        participant = Participant(1)
        participant.set_vote_to_abort(True)

        tx = Transaction(1, ["INSERT INTO table"])
        vote = participant.prepare(tx)

        self.assertEqual(vote, Vote.NO)

    def test_participant_commit(self):
        participant = Participant(1)
        tx = Transaction(1, ["INSERT INTO table"])

        participant.prepare(tx)
        success = participant.commit(1)

        self.assertTrue(success)
        self.assertEqual(participant.state, ParticipantState.COMMITTED)
        self.assertTrue(participant.is_transaction_committed(1))
        self.assertNotIn(1, participant.prepared_transactions)

    def test_participant_abort(self):
        participant = Participant(1)
        tx = Transaction(1, ["INSERT INTO table"])

        participant.prepare(tx)
        success = participant.abort(1)

        self.assertTrue(success)
        self.assertEqual(participant.state, ParticipantState.ABORTED)
        self.assertNotIn(1, participant.prepared_transactions)

    def test_coordinator_all_yes_votes(self):
        coordinator = Coordinator()
        coordinator.add_participant(1)
        coordinator.add_participant(2)

        tx = Transaction(1, ["UPDATE table"])
        coordinator.start_transaction(tx)

        coordinator.receive_vote(1, Vote.YES)
        coordinator.receive_vote(2, Vote.YES)

        self.assertTrue(coordinator.has_all_votes())
        self.assertTrue(coordinator.should_commit())

    def test_coordinator_one_no_vote(self):
        coordinator = Coordinator()
        coordinator.add_participant(1)
        coordinator.add_participant(2)

        tx = Transaction(1, ["UPDATE table"])
        coordinator.start_transaction(tx)

        coordinator.receive_vote(1, Vote.YES)
        coordinator.receive_vote(2, Vote.NO)

        self.assertTrue(coordinator.has_all_votes())
        self.assertFalse(coordinator.should_commit())

    def test_coordinator_decide_commit(self):
        coordinator = Coordinator()
        coordinator.add_participant(1)
        coordinator.add_participant(2)

        tx = Transaction(1, ["UPDATE table"])
        coordinator.start_transaction(tx)

        coordinator.receive_vote(1, Vote.YES)
        coordinator.receive_vote(2, Vote.YES)

        result = coordinator.decide()
        self.assertEqual(result, TransactionResult.COMMITTED)
        self.assertEqual(coordinator.state, CoordinatorState.COMMITTING)

    def test_coordinator_decide_abort(self):
        coordinator = Coordinator()
        coordinator.add_participant(1)

        tx = Transaction(1, ["UPDATE table"])
        coordinator.start_transaction(tx)

        coordinator.receive_vote(1, Vote.NO)

        result = coordinator.decide()
        self.assertEqual(result, TransactionResult.ABORTED)
        self.assertEqual(coordinator.state, CoordinatorState.ABORTING)

    def test_full_commit_transaction(self):
        system = TwoPhaseCommitSystem()
        system.add_participant(1)
        system.add_participant(2)
        system.add_participant(3)

        result = system.execute_transaction([
            "INSERT INTO users VALUES (1, 'Alice')",
            "INSERT INTO accounts VALUES (1, 100)",
        ])

        self.assertEqual(result, TransactionResult.COMMITTED)
        self.assertEqual(system.get_coordinator_state(), CoordinatorState.COMMITTED)

        # All participants should have committed
        for i in [1, 2, 3]:
            self.assertEqual(system.get_participant_state(i), ParticipantState.COMMITTED)

    def test_abort_due_to_participant_failure(self):
        system = TwoPhaseCommitSystem()
        system.add_participant(1)
        system.add_participant(2)
        system.add_participant(3)

        # Participant 2 will vote to abort
        system.set_participant_vote(2, True)

        result = system.execute_transaction(["INSERT INTO users VALUES (1, 'Bob')"])

        self.assertEqual(result, TransactionResult.ABORTED)
        self.assertEqual(system.get_coordinator_state(), CoordinatorState.ABORTED)

        # All participants should have aborted
        for i in [1, 2, 3]:
            self.assertEqual(system.get_participant_state(i), ParticipantState.ABORTED)

    def test_phase1_prepare(self):
        system = TwoPhaseCommitSystem()
        system.add_participant(1)
        system.add_participant(2)

        tx = Transaction(1, ["UPDATE table"])
        votes = system.phase1_prepare(tx)

        self.assertEqual(len(votes), 2)
        self.assertEqual(votes.get(1), Vote.YES)
        self.assertEqual(votes.get(2), Vote.YES)

    def test_atomicity_all_or_nothing(self):
        system = TwoPhaseCommitSystem()
        system.add_participant(1)
        system.add_participant(2)

        system.set_participant_vote(1, True)  # Will abort

        system.execute_transaction(["UPDATE table"])

        # Neither participant should have committed
        self.assertFalse(system.is_transaction_committed(1))

    def test_multiple_transactions(self):
        system = TwoPhaseCommitSystem()
        system.add_participant(1)
        system.add_participant(2)

        result1 = system.execute_transaction(["TX1"])
        self.assertEqual(result1, TransactionResult.COMMITTED)

        system.reset_all()

        result2 = system.execute_transaction(["TX2"])
        self.assertEqual(result2, TransactionResult.COMMITTED)

    def test_single_participant(self):
        system = TwoPhaseCommitSystem()
        system.add_participant(1)

        result = system.execute_transaction(["SINGLE OP"])
        self.assertEqual(result, TransactionResult.COMMITTED)

    def test_all_participants_vote_no(self):
        system = TwoPhaseCommitSystem()
        system.add_participant(1)
        system.add_participant(2)

        system.set_participant_vote(1, True)
        system.set_participant_vote(2, True)

        result = system.execute_transaction(["UPDATE"])
        self.assertEqual(result, TransactionResult.ABORTED)


if __name__ == '__main__':
    unittest.main()

File: Traditional-Important-Algorithms/exercises/06_distributed_systems/dist10_two_phase_commit.rs
// dist10_two_phase_commit.rs
//
// Two-Phase Commit (2PC) is a distributed algorithm that ensures all participants
// in a distributed transaction either commit or abort together. It's crucial for
// maintaining ACID properties across distributed databases.
//
// Protocol phases:
// 1. Prepare Phase: Coordinator asks all participants to prepare (vote yes/no)
// 2. Commit Phase: If all vote yes, coordinator tells all to commit; otherwise abort
//
// Your task: Implement the Two-Phase Commit protocol with coordinator and participants.
//
// Key concepts:
// - Coordinator: Orchestrates the transaction
// - Participants: Nodes that execute parts of the transaction
// - Prepare: Participant locks resources and votes
// - Commit/Abort: Final decision applied atomically
// - Blocking: Protocol blocks if coordinator fails (limitation of 2PC)

// I AM NOT DONE

use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Vote {
    Yes,
    No,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum ParticipantState {
    Idle,
    Prepared,
    Committed,
    Aborted,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum CoordinatorState {
    Idle,
    Preparing,
    Committing,
    Aborting,
    Committed,
    Aborted,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum TransactionResult {
    Committed,
    Aborted,
    InProgress,
}

#[derive(Debug, Clone)]
pub struct Transaction {
    pub id: usize,
    pub operations: Vec<String>,
}

impl Transaction {
    pub fn new(id: usize, operations: Vec<String>) -> Self {
        Self { id, operations }
    }
}

pub struct Participant {
    pub id: usize,
    pub state: ParticipantState,
    pub prepared_transactions: HashMap<usize, Transaction>,
    pub committed_transactions: HashMap<usize, Transaction>,
    pub vote_to_abort: bool,  // Simulates failures
}

impl Participant {
    pub fn new(id: usize) -> Self {
        Self {
            id,
            state: ParticipantState::Idle,
            prepared_transactions: HashMap::new(),
            committed_transactions: HashMap::new(),
            vote_to_abort: false,
        }
    }

    pub fn prepare(&mut self, transaction: Transaction) -> Vote {
        // TODO: Handle prepare request
        // - If vote_to_abort is true, return Vote::No
        // - Otherwise, store transaction in prepared_transactions
        // - Change state to Prepared
        // - Return Vote::Yes
        todo!()
    }

    pub fn commit(&mut self, transaction_id: usize) -> bool {
        // TODO: Handle commit request
        // - Must be in Prepared state
        // - Move transaction from prepared_transactions to committed_transactions
        // - Change state to Committed
        // - Return true if successful, false otherwise
        todo!()
    }

    pub fn abort(&mut self, transaction_id: usize) -> bool {
        // TODO: Handle abort request
        // - Remove transaction from prepared_transactions
        // - Change state to Aborted
        // - Return true
        todo!()
    }

    pub fn set_vote_to_abort(&mut self, should_abort: bool) {
        self.vote_to_abort = should_abort;
    }

    pub fn is_transaction_committed(&self, transaction_id: usize) -> bool {
        self.committed_transactions.contains_key(&transaction_id)
    }

    pub fn reset(&mut self) {
        self.state = ParticipantState::Idle;
    }
}

pub struct Coordinator {
    pub state: CoordinatorState,
    pub transaction: Option<Transaction>,
    pub votes: HashMap<usize, Vote>,
    pub participants: Vec<usize>,
}

impl Coordinator {
    pub fn new() -> Self {
        Self {
            state: CoordinatorState::Idle,
            transaction: None,
            votes: HashMap::new(),
            participants: Vec::new(),
        }
    }

    pub fn add_participant(&mut self, participant_id: usize) {
        self.participants.push(participant_id);
    }

    pub fn start_transaction(&mut self, transaction: Transaction) {
        // TODO: Start a new transaction
        // - Store the transaction
        // - Clear votes
        // - Change state to Preparing
        todo!()
    }

    pub fn receive_vote(&mut self, participant_id: usize, vote: Vote) {
        // TODO: Record a vote from a participant
        // - Store the vote in votes HashMap
        todo!()
    }

    pub fn has_all_votes(&self) -> bool {
        // TODO: Check if all participants have voted
        // - Compare votes.len() with participants.len()
        todo!()
    }

    pub fn should_commit(&self) -> bool {
        // TODO: Determine if transaction should commit
        // - All votes must be Yes
        // - Return false if any vote is No or not all votes received
        todo!()
    }

    pub fn decide(&mut self) -> TransactionResult {
        // TODO: Make commit/abort decision
        // - Must have all votes
        // - If should_commit(), change state to Committing, return Committed
        // - Otherwise, change state to Aborting, return Aborted
        // - If not all votes, return InProgress
        todo!()
    }

    pub fn finalize_commit(&mut self) {
        self.state = CoordinatorState::Committed;
    }

    pub fn finalize_abort(&mut self) {
        self.state = CoordinatorState::Aborted;
    }

    pub fn reset(&mut self) {
        self.state = CoordinatorState::Idle;
        self.transaction = None;
        self.votes.clear();
    }
}

pub struct TwoPhaseCommitSystem {
    coordinator: Coordinator,
    participants: HashMap<usize, Participant>,
    transaction_counter: usize,
}

impl TwoPhaseCommitSystem {
    pub fn new() -> Self {
        Self {
            coordinator: Coordinator::new(),
            participants: HashMap::new(),
            transaction_counter: 0,
        }
    }

    pub fn add_participant(&mut self, participant_id: usize) {
        self.participants.insert(participant_id, Participant::new(participant_id));
        self.coordinator.add_participant(participant_id);
    }

    pub fn set_participant_vote(&mut self, participant_id: usize, vote_to_abort: bool) {
        if let Some(participant) = self.participants.get_mut(&participant_id) {
            participant.set_vote_to_abort(vote_to_abort);
        }
    }

    pub fn execute_transaction(&mut self, operations: Vec<String>) -> TransactionResult {
        // TODO: Execute a complete 2PC transaction
        //
        // Phase 1: Prepare
        // - Create transaction with next transaction_id
        // - Start transaction on coordinator
        // - Send prepare to all participants
        // - Collect votes and send to coordinator
        //
        // Phase 2: Commit or Abort
        // - Coordinator decides based on votes
        // - If Committed: send commit to all participants
        // - If Aborted: send abort to all participants
        // - Finalize coordinator state
        // - Return result
        todo!()
    }

    pub fn phase1_prepare(&mut self, transaction: Transaction) -> HashMap<usize, Vote> {
        // TODO: Execute Phase 1 (Prepare)
        // - For each participant, call prepare
        // - Collect and return all votes
        todo!()
    }

    pub fn phase2_commit(&mut self, transaction_id: usize) {
        // TODO: Execute Phase 2 Commit
        // - For each participant, call commit
        // - Finalize coordinator
        todo!()
    }

    pub fn phase2_abort(&mut self, transaction_id: usize) {
        // TODO: Execute Phase 2 Abort
        // - For each participant, call abort
        // - Finalize coordinator
        todo!()
    }

    pub fn get_coordinator_state(&self) -> CoordinatorState {
        self.coordinator.state
    }

    pub fn get_participant_state(&self, participant_id: usize) -> Option<ParticipantState> {
        self.participants.get(&participant_id).map(|p| p.state)
    }

    pub fn is_transaction_committed(&self, transaction_id: usize) -> bool {
        // Check if all participants have committed the transaction
        self.participants
            .values()
            .all(|p| p.is_transaction_committed(transaction_id))
    }

    pub fn reset_all(&mut self) {
        self.coordinator.reset();
        for participant in self.participants.values_mut() {
            participant.reset();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_participant_prepare_yes() {
        let mut participant = Participant::new(1);
        let tx = Transaction::new(1, vec!["INSERT INTO table".to_string()]);

        let vote = participant.prepare(tx);

        assert_eq!(vote, Vote::Yes);
        assert_eq!(participant.state, ParticipantState::Prepared);
        assert!(participant.prepared_transactions.contains_key(&1));
    }

    #[test]
    fn test_participant_prepare_no() {
        let mut participant = Participant::new(1);
        participant.set_vote_to_abort(true);

        let tx = Transaction::new(1, vec!["INSERT INTO table".to_string()]);
        let vote = participant.prepare(tx);

        assert_eq!(vote, Vote::No);
    }

    #[test]
    fn test_participant_commit() {
        let mut participant = Participant::new(1);
        let tx = Transaction::new(1, vec!["INSERT INTO table".to_string()]);

        participant.prepare(tx);
        let success = participant.commit(1);

        assert!(success);
        assert_eq!(participant.state, ParticipantState::Committed);
        assert!(participant.is_transaction_committed(1));
        assert!(!participant.prepared_transactions.contains_key(&1));
    }

    #[test]
    fn test_participant_abort() {
        let mut participant = Participant::new(1);
        let tx = Transaction::new(1, vec!["INSERT INTO table".to_string()]);

        participant.prepare(tx);
        let success = participant.abort(1);

        assert!(success);
        assert_eq!(participant.state, ParticipantState::Aborted);
        assert!(!participant.prepared_transactions.contains_key(&1));
    }

    #[test]
    fn test_coordinator_all_yes_votes() {
        let mut coordinator = Coordinator::new();
        coordinator.add_participant(1);
        coordinator.add_participant(2);

        let tx = Transaction::new(1, vec!["UPDATE table".to_string()]);
        coordinator.start_transaction(tx);

        coordinator.receive_vote(1, Vote::Yes);
        coordinator.receive_vote(2, Vote::Yes);

        assert!(coordinator.has_all_votes());
        assert!(coordinator.should_commit());
    }

    #[test]
    fn test_coordinator_one_no_vote() {
        let mut coordinator = Coordinator::new();
        coordinator.add_participant(1);
        coordinator.add_participant(2);

        let tx = Transaction::new(1, vec!["UPDATE table".to_string()]);
        coordinator.start_transaction(tx);

        coordinator.receive_vote(1, Vote::Yes);
        coordinator.receive_vote(2, Vote::No);

        assert!(coordinator.has_all_votes());
        assert!(!coordinator.should_commit());
    }

    #[test]
    fn test_coordinator_decide_commit() {
        let mut coordinator = Coordinator::new();
        coordinator.add_participant(1);
        coordinator.add_participant(2);

        let tx = Transaction::new(1, vec!["UPDATE table".to_string()]);
        coordinator.start_transaction(tx);

        coordinator.receive_vote(1, Vote::Yes);
        coordinator.receive_vote(2, Vote::Yes);

        let result = coordinator.decide();
        assert_eq!(result, TransactionResult::Committed);
        assert_eq!(coordinator.state, CoordinatorState::Committing);
    }

    #[test]
    fn test_coordinator_decide_abort() {
        let mut coordinator = Coordinator::new();
        coordinator.add_participant(1);

        let tx = Transaction::new(1, vec!["UPDATE table".to_string()]);
        coordinator.start_transaction(tx);

        coordinator.receive_vote(1, Vote::No);

        let result = coordinator.decide();
        assert_eq!(result, TransactionResult::Aborted);
        assert_eq!(coordinator.state, CoordinatorState::Aborting);
    }

    #[test]
    fn test_full_commit_transaction() {
        let mut system = TwoPhaseCommitSystem::new();
        system.add_participant(1);
        system.add_participant(2);
        system.add_participant(3);

        let result = system.execute_transaction(vec![
            "INSERT INTO users VALUES (1, 'Alice')".to_string(),
            "INSERT INTO accounts VALUES (1, 100)".to_string(),
        ]);

        assert_eq!(result, TransactionResult::Committed);
        assert_eq!(system.get_coordinator_state(), CoordinatorState::Committed);

        // All participants should have committed
        for i in 1..=3 {
            assert_eq!(system.get_participant_state(i), Some(ParticipantState::Committed));
        }
    }

    #[test]
    fn test_abort_due_to_participant_failure() {
        let mut system = TwoPhaseCommitSystem::new();
        system.add_participant(1);
        system.add_participant(2);
        system.add_participant(3);

        // Participant 2 will vote to abort
        system.set_participant_vote(2, true);

        let result = system.execute_transaction(vec![
            "INSERT INTO users VALUES (1, 'Bob')".to_string(),
        ]);

        assert_eq!(result, TransactionResult::Aborted);
        assert_eq!(system.get_coordinator_state(), CoordinatorState::Aborted);

        // All participants should have aborted
        for i in 1..=3 {
            assert_eq!(system.get_participant_state(i), Some(ParticipantState::Aborted));
        }
    }

    #[test]
    fn test_phase1_prepare() {
        let mut system = TwoPhaseCommitSystem::new();
        system.add_participant(1);
        system.add_participant(2);

        let tx = Transaction::new(1, vec!["UPDATE table".to_string()]);
        let votes = system.phase1_prepare(tx);

        assert_eq!(votes.len(), 2);
        assert_eq!(votes.get(&1), Some(&Vote::Yes));
        assert_eq!(votes.get(&2), Some(&Vote::Yes));
    }

    #[test]
    fn test_atomicity_all_or_nothing() {
        let mut system = TwoPhaseCommitSystem::new();
        system.add_participant(1);
        system.add_participant(2);

        system.set_participant_vote(1, true); // Will abort

        system.execute_transaction(vec!["UPDATE table".to_string()]);

        // Neither participant should have committed
        assert!(!system.is_transaction_committed(1));
    }

    #[test]
    fn test_multiple_transactions() {
        let mut system = TwoPhaseCommitSystem::new();
        system.add_participant(1);
        system.add_participant(2);

        let result1 = system.execute_transaction(vec!["TX1".to_string()]);
        assert_eq!(result1, TransactionResult::Committed);

        system.reset_all();

        let result2 = system.execute_transaction(vec!["TX2".to_string()]);
        assert_eq!(result2, TransactionResult::Committed);
    }

    #[test]
    fn test_single_participant() {
        let mut system = TwoPhaseCommitSystem::new();
        system.add_participant(1);

        let result = system.execute_transaction(vec!["SINGLE OP".to_string()]);
        assert_eq!(result, TransactionResult::Committed);
    }

    #[test]
    fn test_all_participants_vote_no() {
        let mut system = TwoPhaseCommitSystem::new();
        system.add_participant(1);
        system.add_participant(2);

        system.set_participant_vote(1, true);
        system.set_participant_vote(2, true);

        let result = system.execute_transaction(vec!["UPDATE".to_string()]);
        assert_eq!(result, TransactionResult::Aborted);
    }
}
